src/utils.cpp

```diff
@@ -1370,17 +1370,29 @@ std::map<uintptr_t, elf_symbol, std::greater<>> get_symbol_table_for_elf(
 std::vector<int> get_pids_for_program(const std::string &program)
 {
   std::vector<int> pids;
-  auto program_abs = std_filesystem::canonical(program);
-  for (const auto &process : std_filesystem::directory_iterator("/proc"))
+
+  std::regex proc_root_regex("/proc/([0-9]+)/root/.*");
+  std::smatch match;
+
+  if (std::regex_search(program, match, proc_root_regex) && match.size() > 1)
   {
-    std::string filename = process.path().filename().string();
-    if (!std::all_of(filename.begin(), filename.end(), ::isdigit))
-      continue;
-    std::error_code ec;
-    std_filesystem::path pid_program = std_filesystem::read_symlink(
-        process.path() / "exe", ec);
-    if (!ec && program_abs == pid_program)
-      pids.emplace_back(std::stoi(filename));
+    pids.emplace_back(std::stoi(match.str(1)));
+  }
+  else
+  {
+    auto program_abs = std_filesystem::canonical(program);
+
+    for (const auto &process : std_filesystem::directory_iterator("/proc"))
+    {
+      std::string filename = process.path().filename().string();
+      if (!std::all_of(filename.begin(), filename.end(), ::isdigit))
+        continue;
+      std::error_code ec;
+      std_filesystem::path pid_program = std_filesystem::read_symlink(
```

Comment:
Idea: I am not sure what this does for programs which are in a different mount namespace. Maybe also trying `std_filesystem::absolute` next to `std_filesystem::canonical` could work.

Reply:
I don't think absolute is correct either, as the absolute path would simply be the path passed in. Perhaps weakly_canonical would work though. I'll give it a go.

Reply:
nope
```
[  FAILED  ] uprobe.uprobes - attach to probe for executable in a pivot_root'd mount namespace
        Command: ../src//bpftrace -e 'uprobe:/proc/20424/root/uprobe_test_static:function1 { printf("func %s\n", func); exit(); }'
        Unclean exit code: 1
        Output: ERROR: failed to write elf: filesystem error: cannot make canonical path: No such file or directory [/proc/20424/root/uprobe_test_static]\n
        ```
Which is strange, because that path does exist, but maybe the problem is that its reading across a filesystem boundary or something.

A solution could be to just catch the exception and return no pids if it gets a file not found. This method is only used to cache symbols for pids at probe attach time. if it doesn't cache anything, that's no big whoop, they can be resolved at output time.